crates/voicevox_core/src/engine/open_jtalk.rs

```diff
@@ -47,55 +52,69 @@ impl PartialEq for OpenJtalkError {
 pub type Result<T> = std::result::Result<T, OpenJtalkError>;
 
 pub struct OpenJtalk {
-    mecab: ManagedResource<Mecab>,
-    njd: ManagedResource<Njd>,
-    jpcommon: ManagedResource<JpCommon>,
+    mecab: Mutex<ManagedResource<Mecab>>,
+    njd: Mutex<ManagedResource<Njd>>,
+    jpcommon: Mutex<ManagedResource<JpCommon>>,
```

Comment:
その場合 `extract_fullcontext` で lifetimeエラーになるのでこのままにします

Reply:
これでいけるのでは?

```diff
     pub fn extract_fullcontext(&self, text: impl AsRef<str>) -> Result<Vec<String>> {
-        let mut jpcommon = self.jpcommon.lock().unwrap();
-        let mut njd = self.njd.lock().unwrap();
-        let mut mecab = self.mecab.lock().unwrap();
+        let Resources {
+            mecab,
+            njd,
+            jpcommon,
+        } = &mut *self.resources.lock().unwrap();
```

Reply:
voicevox_core_c_api 自体は問題なくビルドできましたが、python api の方でコンパイルエラーになりました。詳細は調べていませんが、`#[pyclass]` が対応できなくなるようです。

```
error[E0277]: `*mut *mut i8` cannot be sent between threads safelython_api                                                                           
   --> crates/voicevox_core_python_api/src/lib.rs:79:1
    |
79  | #[pyclass]
    | ^^^^^^^^^^ `*mut *mut i8` cannot be sent between threads safely
    |
    = help: within `voicevox_core::engine::open_jtalk::Resources`, the trait `Send` is not implemented for `*mut *mut i8`
    = note: required because it appears within the type `open_jtalk_sys::bindings::_Mecab`
    = note: required because it appears within the type `std::option::Option<open_jtalk_sys::bindings::_Mecab>`
    = note: required because it appears within the type `open_jtalk::mecab::Mecab`
    = note: required because it appears within the type `open_jtalk::resource::ManagedResource<open_jtalk::mecab::Mecab>`
    = note: required because it appears within the type `voicevox_core::engine::open_jtalk::Resources`
    = note: required for `std::sync::Mutex<voicevox_core::engine::open_jtalk::Resources>` to implement `Sync`
    = note: required because it appears within the type `voicevox_core::OpenJtalk`
    = note: required for `Arc<voicevox_core::OpenJtalk>` to implement `Send`
note: required because it appears within the type `OpenJtalk`
   --> crates/voicevox_core_python_api/src/lib.rs:81:8
    |
81  | struct OpenJtalk {
    |        ^^^^^^^^^
note: required by a bound in `ThreadCheckerStub`
   --> /Users/suitcase/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.18.1/src/impl_/pyclass.rs:849:33
    |
849 | pub struct ThreadCheckerStub<T: Send>(PhantomData<T>);
    |                                 ^^^^ required by this bound in `ThreadCheckerStub`
    = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Reply:
私が提案しようとしていたのは正確には次の通りです。

<details>
<summary>0001-open_jtalk::ManagedResourceに対するMutexを一つにまとめる.patch</summary>

```patch
From f62c8146d643fb56b483eb2cd56c17d3d6a04316 Mon Sep 17 00:00:00 2001
From: Ryo Yamashita <qryxip@gmail.com>
Date: Sat, 18 Mar 2023 18:41:41 +0900
Subject: [PATCH] =?UTF-8?q?`open=5Fjtalk::ManagedResource`=E3=81=AB?=
 =?UTF-8?q?=E5=AF=BE=E3=81=99=E3=82=8B`Mutex`=E3=82=92=E4=B8=80=E3=81=A4?=
 =?UTF-8?q?=E3=81=AB=E3=81=BE=E3=81=A8=E3=82=81=E3=82=8B?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

---
 crates/voicevox_core/src/engine/open_jtalk.rs | 34 +++++++++++--------
 1 file changed, 20 insertions(+), 14 deletions(-)

diff --git a/crates/voicevox_core/src/engine/open_jtalk.rs b/crates/voicevox_core/src/engine/open_jtalk.rs
index 6c43ad2..5857896 100644
--- a/crates/voicevox_core/src/engine/open_jtalk.rs
+++ b/crates/voicevox_core/src/engine/open_jtalk.rs
@@ -52,24 +52,27 @@ impl PartialEq for OpenJtalkError {
 pub type Result<T> = std::result::Result<T, OpenJtalkError>;
 
 pub struct OpenJtalk {
-    mecab: Mutex<ManagedResource<Mecab>>,
-    njd: Mutex<ManagedResource<Njd>>,
-    jpcommon: Mutex<ManagedResource<JpCommon>>,
+    resources: Mutex<Resources>,
     dict_loaded: bool,
 }
 
-#[allow(unsafe_code)]
-unsafe impl Send for OpenJtalk {}
+struct Resources {
+    mecab: ManagedResource<Mecab>,
+    njd: ManagedResource<Njd>,
+    jpcommon: ManagedResource<JpCommon>,
+}
 
 #[allow(unsafe_code)]
-unsafe impl Sync for OpenJtalk {}
+unsafe impl Send for Resources {}
 
 impl OpenJtalk {
     pub fn new_without_dic() -> Self {
         Self {
-            mecab: Mutex::new(ManagedResource::initialize()),
-            njd: Mutex::new(ManagedResource::initialize()),
-            jpcommon: Mutex::new(ManagedResource::initialize()),
+            resources: Mutex::new(Resources {
+                mecab: ManagedResource::initialize(),
+                njd: ManagedResource::initialize(),
+                jpcommon: ManagedResource::initialize(),
+            }),
             dict_loaded: false,
         }
     }
@@ -83,9 +86,11 @@ impl OpenJtalk {
     }
 
     pub fn extract_fullcontext(&self, text: impl AsRef<str>) -> Result<Vec<String>> {
-        let mut jpcommon = self.jpcommon.lock().unwrap();
-        let mut njd = self.njd.lock().unwrap();
-        let mut mecab = self.mecab.lock().unwrap();
+        let Resources {
+            mecab,
+            njd,
+            jpcommon,
+        } = &mut *self.resources.lock().unwrap();
 
         jpcommon.refresh();
         njd.refresh();
@@ -112,7 +117,7 @@ impl OpenJtalk {
             njd.set_accent_type();
             njd.set_unvoiced_vowel();
             njd.set_long_vowel();
-            jpcommon.njd2jpcommon(&njd);
+            jpcommon.njd2jpcommon(njd);
             jpcommon.make_label();
             jpcommon
                 .get_label_feature_to_iter()
@@ -131,9 +136,10 @@ impl OpenJtalk {
 
     fn load(&mut self, open_jtalk_dict_dir: impl AsRef<Path>) -> Result<()> {
         let result = self
-            .mecab
+            .resources
             .lock()
             .unwrap()
+            .mecab
             .load(open_jtalk_dict_dir.as_ref());
         if result {
             self.dict_loaded = true;
-- 
2.39.2

```

</details>

(patch形式なので、`git am -s`すればそのまま取り込めるはずです)

まず`OpenJtalk`が`Send + Sync`でなければならないというのは前提の上で、`Send`と`Sync`の両方を`unsafe impl`している状態であるのを片方だけにできるのでは?という提案です。

Reply:
なるほどです！ ほぼ似たような変更を試していたのですが、こちらの実装では `Send + Sync` 周りの実装が適切ではありませんでした。

```
#[allow(unsafe_code)]
unsafe impl Send for OpenJtalk {}

#[allow(unsafe_code)]
unsafe impl Sync for OpenJtalk {}
```

を削除し、代わりに

```rust
#[allow(unsafe_code)]
unsafe impl Send for Resources {}
```

を追加するということですね。これで python api の方も問題なくビルドできました。

Reply:
複雑度が少し減るのと、元々 mecab, njd, jpcommon を別々に Mutex で包む意義は薄いと感じていたので、提案の通りに変えてしまっても良いかも、と思いました。

Reply:
↑これはやっていいでしょうか?

Reply:
良いと思います！

Reply:
良いと思います！（できれば @qwerty2501 さんの意見もお聞きしたいですが……）

Reply:
↑しました。